misk/src/test/kotlin/misk/hibernate/ReflectionQueryFactoryTest.kt

```diff
@@ -0,0 +1,178 @@
+    val m99 = NameAndReleaseDate("Rocky 99", null)
+
+    transacter.transaction { session ->
+      session.save(DbMovie(m1.name, m1.releaseDate, clock.instant()))
+      session.save(DbMovie(m2.name, m2.releaseDate, clock.instant()))
+      session.save(DbMovie(m3.name, m3.releaseDate, clock.instant()))
+      session.save(DbMovie(m4.name, m4.releaseDate, clock.instant()))
+      session.save(DbMovie(m5.name, m5.releaseDate, clock.instant()))
+      session.save(DbMovie(m98.name, m98.releaseDate, clock.instant()))
+      session.save(DbMovie(m99.name, m99.releaseDate, clock.instant()))
+
+      assertThat(queryFactory.newQuery<MovieQuery>()
+          .releaseDateLessThan(m3.releaseDate)
+          .listAs<NameAndReleaseDate>(session))
+          .containsExactly(m1, m2)
+
+      assertThat(queryFactory.newQuery<MovieQuery>()
+          .releaseDateLessThanOrEqualTo(m3.releaseDate)
+          .listAs<NameAndReleaseDate>(session))
+          .containsExactly(m1, m2, m3)
+
+      assertThat(queryFactory.newQuery<MovieQuery>()
+          .releaseDateEqualTo(m3.releaseDate)
+          .listAs<NameAndReleaseDate>(session))
+          .containsExactly(m3)
+
+      assertThat(queryFactory.newQuery<MovieQuery>()
+          .releaseDateGreaterThanOrEqualTo(m3.releaseDate)
+          .listAs<NameAndReleaseDate>(session))
+          .containsExactly(m3, m4, m5)
+
+      assertThat(queryFactory.newQuery<MovieQuery>()
+          .releaseDateGreaterThan(m3.releaseDate)
+          .listAs<NameAndReleaseDate>(session))
+          .containsExactly(m4, m5)
+
+      assertThat(queryFactory.newQuery<MovieQuery>()
+          .releaseDateNotEqualTo(m3.releaseDate)
+          .listAs<NameAndReleaseDate>(session))
+          .containsExactly(m1, m2, m4, m5)
+    }
+  }
+
+  /** Comparisons with null always return an empty list. */
+  @Test
+  fun comparisonWithNull() {
+    val m1 = NameAndReleaseDate("Rocky 1", LocalDate.of(2018, 1, 1))
+    val m2 = NameAndReleaseDate("Rocky 2", LocalDate.of(2018, 1, 2))
+    val m3 = NameAndReleaseDate("Rocky 3", LocalDate.of(2018, 1, 3))
+    val m4 = NameAndReleaseDate("Rocky 4", LocalDate.of(2018, 1, 4))
+    val m5 = NameAndReleaseDate("Rocky 5", LocalDate.of(2018, 1, 5))
+    val m98 = NameAndReleaseDate("Rocky 98", null)
+    val m99 = NameAndReleaseDate("Rocky 99", null)
+
+    transacter.transaction { session ->
+      session.save(DbMovie(m1.name, m1.releaseDate, clock.instant()))
+      session.save(DbMovie(m2.name, m2.releaseDate, clock.instant()))
+      session.save(DbMovie(m3.name, m3.releaseDate, clock.instant()))
+      session.save(DbMovie(m4.name, m4.releaseDate, clock.instant()))
+      session.save(DbMovie(m5.name, m5.releaseDate, clock.instant()))
+      session.save(DbMovie(m98.name, m98.releaseDate, clock.instant()))
+      session.save(DbMovie(m99.name, m99.releaseDate, clock.instant()))
+
+      assertThat(queryFactory.newQuery<MovieQuery>()
+          .releaseDateLessThan(null)
```

Comment:
this seems buggy. why allow null values if the user is not explicitly looking for null?

Reply:
Yeah, I’m torn between being consistent with what Hibernate does (nulls aren’t special) and avoiding surprising operations (`release_date = NULL` is different from `release_date IS NULL`).

Reply:
I think we get good behavior “for free” by making calling code opt-into nullable types in their parameters. Most queries won’t declare a `?` on the parameter type:

```
    @Constraint(path = "release_date", operator = Operator.LE)
    fun releaseDateLessThanOrEqualTo(upperBound: LocalDate?): MovieQuery
```